open-apparel-registry/open-apparel-registry
# Use contractual limit tracking

## Overview

Previously, a user's number of API calls was tied to the calendar year,
so they could theoretically make 50 calls on Dec 31 and another 50 on
Jan 1. Once we get into paying customers, this becomes more of a
problem. The API limits are now tied to the contract start date, rather
than the calendar year. For free-tier users, we use the date of their
first successful request as the start of their term.

Connects #1410, #1225

## Testing Instructions

* On `develop`, run `./scripts/manage shell_plus`, then enter:
```
from datetime import date
ApiLimit.objects.create(contributor_id=2, yearly_limit=10) #note the id
ApiLimit.objects.create(contributor_id=3, yearly_limit=10)
ApiLimit.objects.create(contributor_id=4, yearly_limit=10)
for x in range(11):
            RequestLog.objects.create(user_id=2, response_code=200)
            r = RequestLog.objects.create(user_id=3, response_code=200)
	    r.created_at = date(2021, 7, 1)
	    r.save()
ApiBlock.objects.create(contributor_id=4, until=date(2021, 12, 31), active=False, limit=10, actual=11) #note the id
ApiBlock.objects.create(contributor_id=5, until=date(2021, 12, 31), active=False, limit=10, actual=11) #note the id
```
* Switch to this branch and run `./scripts/manage migrate`. The migration should succeed.
* Run `./scripts/manage shell_plus`
```
ApiLimit.objects.get(id=1).period_start_date # id matches first limit created. Date should be today
ApiBlock.objects.get(id=1).until # id matches first block created. Date should be today
ApiBlock.objects.get(id=2).until # id matches second block created. The date should still be 12/31/2021 because this block has no associated ApiLimit
```
* Run `./scripts/manage check_api_limits`. An ApiBlock notification should appear.  Run `./scripts/manage shell_plus`
```
ApiBlock.objects.filter(contributor_id=2) # should exist, and "until" is a year from today
ApiBlock.objects.filter(contributor_id=3) # shouldn’t exist since the RequestLogs are all prior to the period_start_date
```
* Run `./scripts/test` and ensure all tests pass

## Checklist
